go/libraries/doltcore/table/untyped/csv/writer.go

```diff
@@ -97,42 +105,150 @@ func (csvw *CSVWriter) GetSchema() schema.Schema {
 func (csvw *CSVWriter) WriteRow(ctx context.Context, r row.Row) error {
 	allCols := csvw.sch.GetAllCols()
 
-	i := 0
-	colValStrs := make([]string, allCols.Size())
-	err := allCols.Iter(func(tag uint64, col schema.Column) (stop bool, err error) {
+	colValStrs := make([]string, 0, allCols.Size())
+	colIsNull := make([]bool, 0, allCols.Size())
+	_, err := r.IterSchema(csvw.sch, func(tag uint64, val types.Value) (stop bool, err error) {
 		val, ok := r.GetColVal(tag)
-		if ok && !types.IsNull(val) {
-			if val.Kind() == types.StringKind {
-				colValStrs[i] = string(val.(types.String))
-			} else {
-				var err error
-				colValStrs[i], err = types.EncodedValue(ctx, val)
+		if !ok || types.IsNull(val) {
+			colIsNull = append(colIsNull, true)
+			colValStrs = append(colValStrs, "")
+			return false, nil
+		}
 
-				if err != nil {
-					return false, err
-				}
+		colIsNull = append(colIsNull, false)
+		if val.Kind() == types.StringKind {
+			colValStrs = append(colValStrs, string(val.(types.String)))
+		} else {
+			var err error
+			v, err := types.EncodedValue(ctx, val)
+			if err != nil {
+				return false, err
 			}
+			colValStrs = append(colValStrs, v)
 		}
 
-		i++
 		return false, nil
 	})
 
 	if err != nil {
 		return err
 	}
 
-	return csvw.csvw.Write(colValStrs)
+	return csvw.write(colValStrs, colIsNull)
 }
 
 // Close should flush all writes, release resources being held
 func (csvw *CSVWriter) Close(ctx context.Context) error {
-	if csvw.closer != nil {
-		csvw.csvw.Flush()
+	if csvw.wr != nil {
+		_ = csvw.wr.Flush()
 		errCl := csvw.closer.Close()
-		csvw.closer = nil
+		csvw.wr = nil
 		return errCl
 	} else {
 		return errors.New("Already closed.")
 	}
 }
+
+// write is directly copied from csv.Writer.Write() with the addition of the `isNull []bool` parameter
+// this method has been adapted for Dolt's special quoting logic, ie `10,,""` -> (10,NULL,"")
+func (csvw *CSVWriter) write(record []string, isNull []bool) error {
```

Comment:
Small nit: just make this take []*string and pass nils for the nil strings. No need for the second array.